tests/unit/test_speech.py

```diff
@@ -0,0 +1,325 @@
+	CharacterModeCommand,
+	PitchCommand,
+	BeepCommand,
+	LangChangeCommand
+)
+
+
+class Test_getSpellingSpeechAddCharMode(unittest.TestCase):
+	def test_symbolNamesAtStartAndEnd(self):
+		# Spelling ¡hola!
+		seq = (c for c in [
+			'inverted exclamation point',
+			EndUtteranceCommand(),
+			'h',
+			EndUtteranceCommand(),
+			'o',
+			EndUtteranceCommand(),
+			'l',
+			EndUtteranceCommand(),
+			'a',
+			EndUtteranceCommand(),
+			'bang',
+			EndUtteranceCommand()
+		])
+		expected = repr([
+			'inverted exclamation point',
+			EndUtteranceCommand(),
+			CharacterModeCommand(True),
+			'h',
+			EndUtteranceCommand(),
+			'o',
+			EndUtteranceCommand(),
+			'l',
+			EndUtteranceCommand(),
+			'a',
+			EndUtteranceCommand(),
+			CharacterModeCommand(False),
+			'bang',
+			EndUtteranceCommand()
+		])
+		output = _getSpellingSpeechAddCharMode(seq)
+		self.assertEqual(repr(list(output)), expected)
+
+	def test_manySymbolNamesInARow(self):
+		# Spelling a...b
+		seq = (c for c in [
+			'a',
+			EndUtteranceCommand(),
+			'dot',
+			EndUtteranceCommand(),
+			'dot',
+			EndUtteranceCommand(),
+			'dot',
+			EndUtteranceCommand(),
+			'b',
+			EndUtteranceCommand()
+		])
+		expected = repr([
+			CharacterModeCommand(True),
+			'a',
+			EndUtteranceCommand(),
+			CharacterModeCommand(False),
+			'dot',
+			EndUtteranceCommand(),
+			'dot',
+			EndUtteranceCommand(),
+			'dot',
+			EndUtteranceCommand(),
+			CharacterModeCommand(True),
+			'b',
+			EndUtteranceCommand()
+		])
+		output = _getSpellingSpeechAddCharMode(seq)
+		self.assertEqual(repr(list(output)), expected)
+
+
+class Test_getSpellingCharAddCapNotification(unittest.TestCase):
+	def test_noNotifications(self):
+		expected = repr([
+			'A',
+		])
+		output = _getSpellingCharAddCapNotification(
+			speakCharAs='A',
+			sayCapForCapitals=False,
+			capPitchChange=0,
+			beepForCapitals=False,
+		)
+		self.assertEqual(repr(list(output)), expected)
+
+	def test_pitchNotifications(self):
+		expected = repr([
+			PitchCommand(offset=30),
+			'A',
+			PitchCommand()
+		])
+		output = _getSpellingCharAddCapNotification(
+			speakCharAs='A',
+			sayCapForCapitals=False,
+			capPitchChange=30,
+			beepForCapitals=False,
+		)
+		self.assertEqual(repr(list(output)), expected)
+
+	def test_beepNotifications(self):
+		expected = repr([
+			BeepCommand(2000, 50, left=50, right=50),
+			'A',
+		])
+		output = _getSpellingCharAddCapNotification(
+			speakCharAs='A',
+			sayCapForCapitals=False,
+			capPitchChange=0,
+			beepForCapitals=True,
+		)
+		self.assertEqual(repr(list(output)), expected)
+
+	def test_capNotifications(self):
+		expected = repr([
+			'cap ',
+			'A',
+		])
+		output = _getSpellingCharAddCapNotification(
+			speakCharAs='A',
+			sayCapForCapitals=True,
+			capPitchChange=0,
+			beepForCapitals=False,
+		)
+		self.assertEqual(repr(list(output)), expected)
+
+	@unittest.skip("Need to implement patching of _ function.")
+	def test_capNotificationsWithPlaceHolderAfter(self):

```

Comment:
I would use a [setup / teardown,](https://docs.python.org/3/library/unittest.html#class-and-module-fixtures) there is nicer syntax for this in pyTest, but lets stick with what we've got:
``` python
class Test_getSpellingCharAddCapNotification(unittest.TestCase):
	_translationResults: typing.Dict[str, str] = {}
	
	@classmethod
	def translationFunctionFake(cls, s: str)
		if str in cls._translationResults:
			return cls._translationResults[str]
		return cls.originalTranslationFunction(s)

	@classmethod
	def setUpClass(cls):
		global _
		cls.originalTranslationFunction = _
		_ = cls.translationFunctionFake

	@classmethod
	def tearDownClass(cls):
		global _
		_ = cls.originalTranslationFunction 

	...
	Some code later
	...
		def test_capNotificationsWithPlaceHolderAfter(self):
			self._translationResults["cap %"] = "%s cap"
			... etc
```

Reply:
I have modified the unit test implementation according to your suggestion.
But the `test_capNotificationsWithPlaceHolderBefore` test (renamed) still doesn't work. For now I have skipped it to show that the other tests work.

The issue is that the patching of `_` function (gettext translation function) is not working as I would expect. Probably a Python issue rather than a unittest lib issue.

@feerrenrut, could you please help me on this point? Thanks.
Note: the skip decorator has to be removed.

Reply:
I see the issue, you have in several places I typed `str` when I meant `s`. I have a more tidy solution for you now, I have pushed this to your branch, I hope you don't mind.

Reply:
> I see the issue, you have in several places I typed `str` when I meant `s`.

Oh yes, my bad.
Just for curiosity, I spent 5 minutes trying to make it work with 's' instead of 'str' but did not had any success. I finally gave up due to the other solution you implemented.

> I have a more tidy solution for you now, I have pushed this to your branch, I hope you don't mind.

No problem.
I would not have found the gettext specific class trick; I am not very familiar with gettext.
Specifically, after reviewing your commit, the only thing I am not able to explain is why installing an instance of `gettext.NullTranslations' is equivalent to `gettext.install('nvda')`.

If you can clarify it for my curiosity and so that I may be able to explain what stands in my own PR, that would be nice.

In any case, I close this conversation since the point is resolved. Thanks!